### ISSUE TITLE
Ability to resolve multiple instances of the same Rule Type with different config

### ISSUE THREAD

**Issue by snikolayev** [NRules/NRules#65]
Question about resolving dependencies or activating rules at run-time. There is a number of rules that check the same criteria for different values and then action different treatments.

A custom RuleActivator implements the IRuleActivator interface, but the Activate method only takes a Type which is fine in circumstances where there is one rule instance per Type. It doesn't allow multiple instances of the same Type however.

Here's a quick example:

    public class DebtStageTenureRule : Rule
    {
      private DebtStage debtStage;

      private int minimumAccountTenure;

      private Treatment treatment;

      public DebtStageTenureRule(DebtStage debtStage, int minimumAccountTenure, Treatment treatment)
      {
         this.debtStage = debtStage;
         this.minimumAccountTenure = minimumAccountTenure;
         this.treatment = treatment;
      }

      public override void Define()
      {
         Consumer consumer = null;

         When()
            .Match<Consumer>(() => consumer, c => c.DebtStage.Equals(debtStage))
            .Match<Consumer>(c => c.Tenure >= minimumAccountTenure);

         Then()
            .Do(context => consumer.ApplyTreatment(treatment));
      }
    }

Now let's say I want to configure two rules of this Type using Spring Framework IoC.
1) Debt stage of A and MinimumTenure of 300 gives a treatment of TreatmentX.
2) Debt stage of B and MinimumTenure of 500 gives a treatment of TreatmentY.

Would it possible to do this without making DebtStageTenureRule an abstract class and creating two inheriting classes just for the different Types?

I noticed the Name attribute you can place on a Rule but I presume this is just metadata for logging and/or event handler context.

**Comment by peterjrichards** [comment]
Has this enhancement been implemented? If so, please could you update the documentation or provide an example of how to add the instances?

**Comment by snikolayev** [comment]
The key to using multiple rules per rule type is designing a rule such that it can be instantiated with different options, to serve different purposes. For example, the following rule type can be instantiated with different limit values to match different order totals:
```c#
public class LargeTotalRule : Rule
{
    public double Limit { get; }

    public LargeTotalRule(double limit)
    {
        Limit = limit;
    }

    public override void Define()
    {
        Customer customer = null;
        IEnumerable<Order> orders = null;
        double total = 0;

        Name($"TotalOver{Limit}Rule");

        When()
            .Match<Customer>(() => customer, c => c.IsPreferred)
            .Query(() => orders, x => x
                .Match<Order>(
                    o => o.Customer == customer,
                    o => o.IsOpen)
                .Collect())
            .Let(() => total, () => orders.Sum(x => x.Amount))
            .Having(() => total > Limit);

        Then()
            .Do(ctx => Console.WriteLine("Customer {0} has orders over ${1} total", customer.Name, Limit));
    }
}
```

When using an Autofac-based rule activator, we can now register different instances of this rule, with different options, and the rule activator will return all these instances when activating the corresponding rule type.

```c#
var builder = new ContainerBuilder();
builder.Register(x => new LargeTotalRule(100)).AsSelf();
builder.Register(x => new LargeTotalRule(50)).AsSelf();
var container = builder.Build();

var repository = new RuleRepository();
repository.Activator = new AutofacRuleActivator(container);
repository.Load(x => x.From(Assembly.GetExecutingAssembly()));
```

Now we have two rules loaded into the `RuleRepository` - TotalOver100Rule and TotalOver50Rule.

**Comment by snikolayev** [comment]
@peterjrichards and yes, it has been  implemented and released long ago, with version 0.4.1.
